File: yukicoder/100-199/177.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <queue>
#define REP(i, a, b) for (int i = (int)(a); i < (int)(b); i++)
#define rep(i, a) REP(i, 0, a)

using namespace std;

template <class T>
class MaxFlowDinic {
 public:
    // you may need to change INF
    static const T INF = 2147483647; //INT_MAX
    int V;
    struct edge {
        int to, rev;
        T cap;
    };
    vector<vector<edge>> G;
    vector<int> level;
    vector<int> iter;

    MaxFlowDinic() {};
    MaxFlowDinic(int v) {
        V = v;
        G.resize(V); level.resize(V); iter.resize(V);
        for (int i = 0; i < V; i++) G[i].clear();
    }

    void add(int from, int to, T cap) {
        G[from].push_back(edge{to, (int)G[to].size(), cap});
        G[to].push_back(edge{from, (int)G[from].size() - 1, 0});
    }

    void add_dual(int from, int to, T cap) {
        G[from].push_back(edge{to, (int)G[to].size(), cap});
        G[to].push_back(edge{from, (int)G[from].size() - 1, cap});
    }

    void bfs(int s) {
        fill(level.begin(), level.end(), -1);
        queue<int> que;
        level[s] = 0;
        que.push(s);
        while (!que.empty()) {
            int v = que.front(); que.pop();
            for (int i = 0; i < G[v].size(); i++) {
                edge &e = G[v][i];
                if (e.cap > 0 && level[e.to] < 0) {
                    level[e.to] = level[v] + 1;
                    que.push(e.to);
                }
            }
        }
    }

    T dfs(int v, int t, T f) {
        if (v == t) return f;
        for (int &i = iter[v]; i < G[v].size(); i++) {
            edge &e = G[v][i];
            if (e.cap > 0 && level[v] < level[e.to]) {
                T d = dfs(e.to, t, min(f, e.cap));
                if (d > 0) {
                    e.cap -= d;
                    G[e.to][e.rev].cap += d;
                    return d;
                }
            }
        }
        return 0;
    }


    T max_flow(int s, int t) {
        T flow = 0;
        while (1) {
            bfs(s);
            if (level[t] < 0) return flow;
            fill(iter.begin(), iter.end(), 0);
            T f;
            while ((f = dfs(s, t, INF)) > 0) {
                flow += f;
            }
        }
    }
};


int W, N, M, Q, X;
int J[51], C[51];
int NG[51][51];

int main(void) {
    cin >> W >> N;
    rep(i, N) cin >> J[i];
    cin >> M;
    rep(i, M) cin >> C[i];

    MaxFlowDinic<int> mf(1000);
    rep(i, N) mf.add(0, i + 1, J[i]);
    rep(i, M) mf.add(100 + i, 200, C[i]);
    rep(i, M) {
        cin >> Q;
        rep(j, Q) {
            cin >> X;
            NG[X - 1][i] = 1;
        }
        rep(j, N) {
            if (NG[j][i] == 0) mf.add(j + 1, 100 + i, 1 << 28);
        }
    }


    int cut = mf.max_flow(0, 200);
    if (cut < W) cout << "BANSAKUTSUKITA" << endl;
    else    cout << "SHIROBAKO" << endl;

    return 0;
}

File: yukicoder/100-199/177.py
class MaxFlowDinic:
    class edge:
        def __init__(self, to, rev, cap):
            self.to, self.rev, self.cap = to, rev, cap

    def __init__(self, v):
        self.V = v
        self.G = [[] for i in xrange(self.V)]

    def add(self, frm, to, cap):
        self.G[frm].append(self.edge(to, len(self.G[to]), cap))
        self.G[to].append(self.edge(frm, len(self.G[frm]) - 1, 0))

    def add_dual(self, frm, to, cap):
        self.G[frm].append(self.edge(to, len(self.G[to]), cap))
        self.G[to].append(self.edge(frm, len(self.G[frm]) - 1, cap))

    def bfs(self, s):
        self.level = [-1] * self.V
        self.level[s] = 0
        que = []
        que.append(s)
        while (que):
            v = que.pop(0)
            for i in xrange(len(self.G[v])):
                e = self.G[v][i];
                if e.cap > 0 and self.level[e.to] < 0:
                    self.level[e.to] = self.level[v] + 1;
                    que.append(e.to)


    def dfs(self, v, t, f):
        if v == t: return f
        for i in xrange(self.itr[v], len(self.G[v])):
            self.itr[v] = i
            e = self.G[v][i]
            if e.cap > 0 and self.level[v] < self.level[e.to]:
                d = self.dfs(e.to, t, min(f, e.cap))
                if d > 0:
                    e.cap -= d
                    self.G[e.to][e.rev].cap += d
                    return d
        return 0

    def max_flow(self, s, t):
        inf = float("inf")
        flow = 0
        while 1:
            self.bfs(s)
            if self.level[t] < 0: return flow

            self.itr = [0] * self.V
            f = self.dfs(s, t, inf)
            while f > 0:
                flow += f
                f = self.dfs(s, t, inf)


W, N = int(raw_input()), int(raw_input())
J = map(int, raw_input().split())
M = int(raw_input())
C = map(int, raw_input().split())

mf = MaxFlowDinic(201)

for i in xrange(N):
    mf.add(0, i + 1, J[i])

for i in xrange(M):
    mf.add(100 + i, 200, C[i])

NG = [[False] * M for i in xrange(N)]
for i in xrange(M):
    QX = map(int, raw_input().split())
    for j in QX[1:]:
        NG[j - 1][i] = True
    for j in xrange(N):
        if not NG[j][i]:
            mf.add(j + 1, 100 + i, float("inf"))

cut = mf.max_flow(0, 200)
print "SHIROBAKO" if cut >= W else "BANSAKUTSUKITA"
